cool-core/src/main/java/com/nus/cool/core/cohort/refactor/olapSelect/olapAggregation.java

```diff
@@ -0,0 +1,329 @@
+      for (Map.Entry<String, BitSet> entry : this.group.entrySet()) {
+        for (Map.Entry<String, BitSet> nextEntry : next.entrySet()) {
+          String groupName = entry.getKey() + "|" + nextEntry.getKey();
+          BitSet bs = (BitSet) entry.getValue().clone();
+          bs.and(nextEntry.getValue());
+          merged.put(groupName, bs);
+        }
+      }
+      this.group = merged;
+    }
+  }
+
+  /**
+   * groupBy aggregation instance
+   *
+   * @param bs            bitMap, true means the record in time range
+   * @param groupbyFields filed ot be group by
+   * @param metaChunk     current metaChunk
+   * @param dataChunk     current metaChunk
+   */
+  public void groupBy(BitSet bs, List<String> groupbyFields, MetaChunkRS metaChunk,
+                      ChunkRS dataChunk, granularityType GroupFields_granularity) {
+
+    System.out.println("Init OLAP query timeRanges, key = " + timeRange + ", matched records = " +
+        bs.cardinality() + ", total size of BS = " + bs.size());
+
+    this.timeRange = timeRange;
+    this.dataChunk = dataChunk;
+    this.metaChunk = metaChunk;
+    if (groupbyFields == null) {
+      this.group.put("all", bs);
+      return;
+    }
+    // group By multiple fields,
+    for (String groupbyField : groupbyFields) {
+      MetaFieldRS metaField = metaChunk.getMetaField(groupbyField);
+      FieldRS dataField = dataChunk.getField(groupbyField);
+      switch (dataField.getFieldType()) {
+        case UserKey:
+        case Action:
+        case Segment:
+          group(dataField, bs, metaField, GroupType.STRING);
+          break;
+        case Metric:
+          group(dataField, bs, metaField, GroupType.NUMERIC);
+          break;
+        case ActionTime:
+          group(dataField, bs, GroupFields_granularity);
+          break;
+        default:
+          throw new UnsupportedOperationException(
+              "Unsupport field type: " + dataField.getFieldType());
+      }
+    }
+    mergeGroups();
+  }
+
+  public ArrayList<OlapRet> process(Aggregation aggregation, HashSet<String> projectedSchemaSet) {
+
+    // init projectedTuple.
+    ProjectedTuple tuple = new ProjectedTuple(projectedSchemaSet);
+
+    // 1. get the field type and data
+    String fieldName = aggregation.getFieldName();
+    FieldType fieldType = this.metaChunk.getMetaField(fieldName).getFieldType();
+
+    FieldRS field = this.dataChunk.getField(fieldName);
+    InputVector valueVec = field.getValueVector();
+
+    // 2. init the Aggregate function
+    Map<AggregateType, AggregateFunc > aggMap = new HashMap<>();
+    for (AggregateType operator: aggregation.getOperators()) {
+      if (!checkOperatorIllegal(fieldType, operator)) {
+        throw new IllegalArgumentException(fieldName + " can not process " + operator);
+      }
+      AggregateFunc aggregator = AggregateFactory.generateAggregate(operator, fieldName);
+      aggMap.put(operator, aggregator);
+    }
+
+    // 3. init the result
```

Comment:
Here use RetUnit array directly, so CohortRet is not used?

Reply:
OLAPRET is for after processing